DataFormats/Common/interface/DetSetVectorNew.h

```diff
@@ -470,27 +649,46 @@ namespace edmNew {
   }
 
   template<typename T>
-  inline void DetSetVector<T>::updateImpl(Item & item)  {
+  inline void DetSetVector<T>::updateImpl(Item & item) {
     // no getter or already updated
+/*
     if (getter.empty()) assert(item.offset>=0);
     if (item.offset!=-1 || getter.empty() ) return;
     item.offset = int(m_data.size());
     FastFiller ff(*this,item,true);
     (*boost::any_cast<std::shared_ptr<Getter> >(&getter))->fill(ff);
+*/
+    if (getter.empty()) { assert(item.offset>=0); return;}
+#ifdef USE_ATOMIC
+    int expected = -1;
+    if (item.offset.compare_exchange_strong(expected,-2,std::memory_order_acq_rel)) {
+      assert(item.offset==-2); 
+      {
+	TSFastFiller ff(*this,item);
+	(*boost::any_cast<std::shared_ptr<Getter> >(&getter))->fill(ff);
+      }
+      assert(item.offset>=0);
+    }
+#endif
   }
-
-
  
   
   template<typename T>
   inline void DetSet<T>::set(DetSetVector<T> const & icont,
 			     typename Container::Item const & item, bool update) {
-    if (update) {
-      icont.update(item);
-      assert(item.offset>=0);
-    }
-    m_id=item.id; 
+#ifdef USE_ATOMIC
+    // if an item is being updated we wait
+    if (update)icont.update(item);
+    while(item.offset.load(std::memory_order_acquire)<-1) nanosleep(0,0);
```

Comment:
When would you have the case of ```update==false``` but ```item.offset <-1```? Seems like ```update``` should have the wait until the value is set based on if the ```compare_exchange_*``` succeeded or failed.